src/AggregateFunctions/Streaming/CountedValueMap.h

```diff
@@ -1,52 +1,85 @@
 #pragma once
 
+#include <base/StringRef.h>
+#include <Common/ArenaUtils.h>
+#include <Common/ArenaWithFreeLists.h>
+
 #include <absl/container/btree_map.h>
 
+#include <map>
 #include <stdexcept>
 
 namespace DB
 {
 namespace Streaming
 {
+template <typename T>
+struct CountedValueArena
+{
+    CountedValueArena() = default;
+    T emplace(T key) { return key; }
+    void free(const T & /*key*/) { }
+};
+
+/*
+ * Specialized storage for StringRef with a freelist arena->
+ * Keys of this type that are retained on insertion must be serialized into local storage,
+ * otherwise the reference would be invalid after the processed block is released.
+ */
+template <>
+struct CountedValueArena<StringRef>
+{
+    CountedValueArena() = default;
+    StringRef emplace(StringRef key) { return copyStringInArena(arena, key); }
+
+    void free(StringRef key)
+    {
+        if (key.data)
+            arena.free(const_cast<char *>(key.data), key.size);
+    }
+
+private:
+    ArenaWithFreeLists arena;
+};
+
 /// CountedValueMap maintain count for each key with maximum capacity
 /// When capacity hits the max capacity threshold, it will delete
 /// the minimum / maximum key in the map to maintain the capacity constrain
-template <typename T, bool maximum>
+template <typename T, bool maximum, typename KeyCompare = void>
 class CountedValueMap
 {
 public:
-    explicit CountedValueMap(int64_t max_size_) : max_size(max_size_) { }
+    /// If no provide KeyCompare, we use a default compare (maxiumum - greater<T>, minimum - less<T>)
+    using Compare = std::conditional_t<std::is_void_v<KeyCompare>, std::conditional_t<maximum, std::greater<T>, std::less<T>>, KeyCompare>;
+
+    /// NOTE: If the `Compare` is nothrow copy constructible, we prefer to use absl::btree_map, otherwise use std::map
```

Comment:
Can we just use absl ? Or no throw required by absl::btree_map ?

Reply:
yes, absl::btree_map  requires `Compare` is nothrow copy constructible, so i updated the comment:
```sql
/// Generally we prefer to use absl::btree_map, but it requires `Compare` is nothrow copy constructible, so if not, then we use std::map
```

Reply:
std::greater<T> etc is not noexcept ?